agent/lib_mongodb.c

```diff
@@ -173,7 +175,205 @@ NR_PHP_WRAPPER(nr_mongodb_operation) {
 }
 NR_PHP_WRAPPER_END
 
-void nr_mongodb_enable(TSRMLS_D) {
+#else
+
+NR_PHP_WRAPPER(nr_mongodb_operation_before) {
+    (void)wraprec;
+    nr_segment_t* segment;
+    segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+    segment->wraprec = auto_segment->wraprec;
+}
+NR_PHP_WRAPPER_END
+
+NR_PHP_WRAPPER(nr_mongodb_operation_after) {
+  const char* this_klass = "MongoDB\\Operation\\Executable";
+  zval* collection = NULL;
+  zval* database = NULL;
+  zval* server = NULL;
+  zval* this_var = NULL;
+  nr_datastore_instance_t instance = {
+      .host = NULL,
+      .port_path_or_id = NULL,
+      .database_name = NULL,
+  };
+  nr_segment_datastore_params_t params = {
+    .datastore = {
+      .type = NR_DATASTORE_MONGODB,
+    },
+    .operation = nr_strdup (wraprec->extra),
+    .instance = &instance,
+    .callbacks = {
+      .backtrace = nr_php_backtrace_callback,
+    },
+  };
+
+  /*
+   * We check for the interface all Collection operations extend, rather than
+   * their specific class. Not all operations have the properties we need but
+   * the ones we hook do (as of mongo-php-library v.1.1).
+   */
+  this_var = nr_php_scope_get(NR_EXECUTE_ORIG_ARGS);
+  if (!nr_php_object_instanceof_class(this_var, this_klass)) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "%s: operation is not %s", __func__,
+                     this_klass);
+    goto leave;
+  }
+
+  collection
+      = nr_php_get_zval_object_property(this_var, "collectionName");
+  if (nr_php_is_zval_valid_string(collection)) {
+    params.collection = Z_STRVAL_P(collection);
+  }
+
+  database
+      = nr_php_get_zval_object_property(this_var, "databaseName");
+  if (nr_php_is_zval_valid_string(database)) {
+    instance.database_name = Z_STRVAL_P(database);
+  }
+
+  server = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS);
+  nr_mongodb_get_host_and_port_path_or_id(server, &instance.host,
+                                          &instance.port_path_or_id);
+
+  nr_segment_datastore_end(&auto_segment, &params);
+
+leave:
+  nr_php_arg_release(&server);
+  nr_php_scope_release(&this_var);
+  nr_free(instance.host);
+  nr_free(instance.port_path_or_id);
+  nr_free(params.operation);
+}
+NR_PHP_WRAPPER_END
+
+#endif /* OAPI */
+
+void nr_mongodb_enable() {
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
+    && !defined OVERWRITE_ZEND_EXECUTE_DATA
+
+    nr_php_wrap_user_function_before_after_clean_extra(
+        NR_PSTR("MongoDB\\Operation\\Aggregate::execute"),
+        nr_mongodb_operation_before,
+        nr_mongodb_operation_after,
+        NULL,
```

Comment:
Because "before" is always starting a segment, we will have dangling segments if "after" is never called (for example when an exception occurred). The remedies for this are either 1) create a "clean" function that closes the segment or 2) add the "after" function as the "clean" argument. This means that the "after" function needs to verify the return value and any arguments it uses.

I recommend 2 because that needs to be done anyway for https://github.com/newrelic/newrelic-php-agent/pull/866

Reply:
fixed in 0b911e485feb13860c0eb6db8bfd458405272b8d